### PR TITLE
Fix #1316: list support

### LINKED ISSUE #1316: Homogenously-typed lists in nopython mode
(This issue is mainly for discussion.  I don't have a concrete proposal for how this should work.)

Lists are useful (as are list comprehensions), and with NRT we could support mutating list contents and length.  However, to support lists in nopython mode we would need the elements of a given list to have a single type, and we would need to be able to infer this type.

We can certainly determine the element type for a non-empty list literal (like `[4, 3]`), but it would be less obvious for lists that start as an empty list, then have values appended to them later.  I don't know if our current type unification algorithm can handle that.

In cases where we cannot guarantee a consistent type for the list elements (for example, if two incompatible types are appended to the list in the function body), we would have to raise an exception in nopython mode.

There is the additional problem that list arguments need to appear to be passed by reference, which basically means that we can only reasonably support:

 * Lists from Python passed as arguments (i.e. lists with a Python interpreter reference): read-only
 * Lists created within nopython (i.e. lists with no Python interpreter references) mode: read-write

I think these are acceptable limitations, but we will need very clear exception messages to explain what is going on.

(A solution for lists would also open the possibility of homogenously-typed dictionaries, where a dictionary instance would have a fixed key type and a fixed value type.)

### PR DESCRIPTION
**Pull request by pitrou**
Most list features are implemented, except:
* unboxing support (CPython -> Numba) (mutations to the unboxed list wouldn't be reflected to CPython, which would be confusing)
* in-place operators (need further changes to typeinfer and lowering; not complicated, but still deserves separate attention)
* list.sort() (too much additional work for this already large PR)

Needs an up-to-date llvmlite, because of https://github.com/numba/llvmlite/pull/86

### CODE REVIEW

**Comment by seibert** [comment]
Regarding `list.sort()`: I see that CPython uses a non-recursive mergesort for lists.  Would it make sense for us to do the same?  (Not saying that we need to support `list.sort()` in this PR, though.)

**Comment by pitrou** [comment]
Actually, the sort algorithm is called timsort and has very good performance properties. See https://en.wikipedia.org/wiki/Timsort and also https://hg.python.org/cpython/file/tip/Objects/listsort.txt

We should definitely copy it, or some of it depending on complexity (perhaps leave out some heuristics and special cases). I thought the implementation was recursive, but if it's iterative then all the better :)

**Comment by seibert** [comment]
Does this PR implement a way to create a numpy array from a list?  (Ex: `np.array(my_list)`)

**Comment by pitrou** [comment]
No, it doesn't. That would be for a separate PR.
(neither does it allow for `tuple(list)`, for instance)

**Comment by seibert** [comment]
OK, fair enough.  This PR is sufficiently long as it is.  Let's get this reviewed, and then move on to "Advanced Lists".

**Comment by sklam** [comment]
`restypes` at https://github.com/pitrou/numba/blob/listobj/numba/typeinfer.py#L277 is unused

File: numba/datamodel/models.py [hunk 1]
```diff
@@ -208,9 +222,22 @@ def __init__(self, dmm, fe_type):
...
+
+    def has_nrt_meminfo(self):
+        return True
+
     def get_nrt_meminfo(self, builder, value):
-        if self._fe_type == types.meminfo_pointer:
-            return value
+        for tp in self.inner_types():
+            if self._dmm.lookup(tp).has_nrt_meminfo():
+                raise NotImplementedError(
```
Comment (sklam) [inline numba/datamodel/models.py]: Did you encounter something that require nested MemInfo?  I imagine other managed object to exist as reference (CPointer) inside other MemInfo.

Comment (pitrou) [inline numba/datamodel/models.py]: I tried to play with nested lists :)

**Comment by pitrou** [comment]
I pushed fixes to Siu's remarks.

**Comment by seibert** [comment]
Alright, here we go...

**Pull request merged**